File: VinitKumar/Problems/Milestone1/Problem10.cs
using System;


using System;
using VinitKumar.Utilities;

/*
 *  [102. Binary Tree Level Order Traversal](https://leetcode.com/problems/binary-tree-level-order-traversal/description/)
 */

namespace VinitKumar.Problems.Milestone1
{
	public class Problem10
	{
        IList<IList<int>>? ansR;
        private IList<IList<int>> LevelOrderIterative(TreeNode? root)
        {
            IList<IList<int>> ans = new List<IList<int>>();
            if (root == null)
            {
                return ans;
            }
            Queue<TreeNode> q = new Queue<TreeNode>();
            q.Enqueue(root);
            while (q.Count > 0)
            {
                int qsz = q.Count;
                List<int> tlist = new List<int>();
                while (qsz > 0)
                {
                    TreeNode crnt = q.Dequeue();
                    tlist.Add(crnt.val);

                    if (crnt.left != null)
                    {
                        q.Enqueue(crnt.left);
                    }
                    if (crnt.right != null)
                    {
                        q.Enqueue(crnt.right);
                    }
                    qsz -= 1;
                }
                ans.Add(tlist);
            }
            return ans;
        }
        
        private void LevelOrderRecursive(TreeNode? root, int level)
        {
            if (root == null)
            {
                return;
            }

            if (ansR?.Count == level)
            {
                ansR.Add(new List<int>());
            }

            ansR?[level].Add(root.val);

            LevelOrderRecursive(root?.left, level + 1);
            LevelOrderRecursive(root?.right, level + 1);
        }
        public IList<IList<int>> LevelOrder(TreeNode root)
        {
            // return LevelOrderIterative(root);

            ansR = new List<IList<int>>();
            LevelOrderRecursive(root, 0);
            return ansR;
        }

    }
}

File: VishnuThakral/Milestone1/Problem10.py
#Milestone 1Problem 10:Leetcode number 102(https://leetcode.com/problems/binary-tree-level-order-traversal/description/)
# Time: O(N), Space O(N)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        queue = []
        if root:
            seen = 0
            queue.append([root, 0])
            while queue:
                node, lvl = queue.pop(0)
                if seen==0 or seen<lvl:
                    res.append([node.val])
                    seen = lvl
                else:
                    res[lvl].append(node.val)
                if node.left:
                    queue.append([node.left,lvl+1])
                if node.right:
                    queue.append([node.right,lvl+1])
        return res